plugins/languages/treesitter/treesitter.nix

```
@@ -6,210 +6,264 @@
+      default = true;
+      description = "Whether to enable tree-sitter based folding.";
+    };
 
-          custom_captures = if (cfg.customCaptures != { }) then cfg.customCaptures else null;
-        };
+    gccPackage = helpers.mkPackageOption {
+      name = "gcc";
+      default = pkgs.gcc;
+      description = ''
+        Which package (if any) to be added as the GCC compiler.
+        This is required to build grammars if you are not using `nixGrammars`.
+        To disable the installation of GCC, set this option to `null`.
+      '';
+    };
 
-        incremental_selection =
-          if cfg.incrementalSelection.enable then
-            {
-              enable = true;
-              keymaps = {
-                init_selection = cfg.incrementalSelection.keymaps.initSelection;
-                node_incremental = cfg.incrementalSelection.keymaps.nodeIncremental;
-                scope_incremental = cfg.incrementalSelection.keymaps.scopeIncremental;
-                node_decremental = cfg.incrementalSelection.keymaps.nodeDecremental;
-              };
-            }
-          else
-            null;
-
-        indent = if cfg.indent then { enable = true; } else null;
-
-        ensure_installed = if cfg.nixGrammars then [ ] else cfg.ensureInstalled;
-        ignore_install = cfg.ignoreInstall;
-        parser_install_dir = cfg.parserInstallDir;
-      } // cfg.moduleConfig;
-    in
-    mkIf cfg.enable {
-      extraConfigLua =
-        (optionalString (cfg.parserInstallDir != null) ''
-          vim.opt.runtimepath:append("${cfg.parserInstallDir}")
-        '')
-        + ''
-          require('nvim-treesitter.configs').setup(${helpers.toLuaObject tsOptions})
-        ''
-        + (optionalString (cfg.languageRegister != { }) ''
-          __parserFiletypeMappings = ${helpers.toLuaObject cfg.languageRegister}
-
-          for parser_name, ft in pairs(__parserFiletypeMappings) do
-            require('vim.treesitter.language').register(parser_name, ft)
-          end
-        '');
-
-      extraFiles = mkIf cfg.nixvimInjections {
-        "queries/nix/injections.scm" = ''
-          ;; extends
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @lua)
-              (indented_string_expression (string_fragment) @lua)
-            ]
-            (#match? @_path "^extraConfigLua(Pre|Post)?$"))
-
-          (binding
-            attrpath: (attrpath (identifier) @_path)
-            expression: [
-              (string_expression (string_fragment) @vim)
-              (indented_string_expression (string_fragment) @vim)
-            ]
-            (#match? @_path "^extraConfigVim(Pre|Post)?$"))
+    grammarPackages = mkOption {
+      type = with types; listOf package;
+      default = config.plugins.treesitter.package.passthru.allGrammars;
+      defaultText = literalExpression "pkgs.vimPlugins.nvim-treesitter.passthru.allGrammars";
```

```suggestion
      default = config.plugins.treesitter.package.passthru.allGrammars or [ ];
      defaultText = literalExpression "config.plugins.treesitter.package.passthru.allGrammars";
```

Not sure, but I _think_ the default may still be evaluated even when some of those assertions fail. If so, we should add `or [ ]` to the value.

Maybe add a test for a null `plugins.treesitter.package` or play around with this in another nixvim config that has `flake lock --override-input` pointing to your branch?

Also, we should update the defaultText to match (don't need the `or [ ]` there though, as it would never apply in real usage)

---

Just setting up a test for a null package doesn't seem like it's even accepted. 
```nix
    plugins.treesitter = {
      enable = true;

      package = null;
    };
```

```bash
error: A definition for option `plugins.treesitter.package' is not of type `package'. Definition values:
       - In `<unknown-file>': null
```

---

Interesting. mkNeovimPlugin [is using](https://github.com/nix-community/nixvim/blob/843fb302ebaa2559b7c50ad69ba9a00ae1a5836d/lib/neovim-plugin.nix#L93) `mkPluginPackageOption`, which is [apparently not nullable](https://github.com/nix-community/nixvim/blob/843fb302ebaa2559b7c50ad69ba9a00ae1a5836d/lib/options.nix#L333-L339).

@GaetanLepage is this what we want? It was my interpretation that our package options should be nullable.

---

I think for now let's just trust that `cfg.package` is never null, so remove any assertions (etc) related to it being nullable.

If we change `mkPluginPackageOption` in the future we'll have to consider how it affects plugins like this one.